components/datalogger/datalogger_utils.py

```diff
@@ -230,3 +232,252 @@ def FindAuxControlDataComponentInResourceConfiguration(a_resource_configuration)
         except KeyError:
             pass
     return None
+
+# Write a variable list of "objects" each of which has a variable list of "items" to a serialised string that can be output to the CSV.
+# The items will each have a "title" property that identifies the CSV header (effectively the column) that the data is to be associated
+# with and written to. An example of an object may be a point of interest on the left of a machine blade or bucket. That point of interest
+# will then have items # representing the x, y and z coordinate for that point of interest with a title of a form similar to "blade left [x]",
+# "blade left [y]" and "blade left [z]". This function would hence produce 3 outputs for that object before moving to the next. This approach
+# allows objects to be defined outside of this object with no knowledge required of the structure of what's being serialised. For
+# example another object may represent a single point in space represented by a WGS84 coordinate. That object's item list then may contain
+# 3 items called "lat", "lon" and "height". Objects with shorter or longer item lists are also possible making this function flexible.
+# The algorithm uses the title_list to managed where in the CSV output string each item is written so that data aligns with the comma
+# separated titles when they're eventually written to file. State is written in a similar way but at fixed (known) column offsets.
+#
+def OutputLineObjects(a_file_ptr, a_machine_type, a_replicate, a_assets_dict, aux_control_data_dict, a_object_list, a_header_list, a_state={}):
+    ac_uuid = a_replicate["data"]["ac_uuid"]
+    machine_name = "-"
+    device_id = "-"
+    operator_id = "-"
+    task_id = "-"
+    delay_id = "-"
+    surface_name = "-"
+
+    try:
+        operator_id = a_state["topcon.rdm.list"]["operator"]["value"]
+    except KeyError:
+        logging.debug("No Operator state found.")
+
+    try:
+        delay_id = a_state["topcon.rdm.list"]["delay"]["value"]
+    except KeyError:
+        logging.debug("No Delay state found.")
+
+    try:
+        task_id = a_state["topcon.task"]["id"]["value"]
+    except KeyError:
+        logging.debug("No Task state found.")
+
+    try:
+        surface_name = a_state["topcon.task"]["surface_name"]["value"]
+    except KeyError:
+        logging.debug("No Surface state found.")
+
+
+    for i, val in enumerate(a_assets_dict[ac_uuid]["signatures"]):
+        if val["asset_urn"].startswith("urn:X-topcon:machine"):
```

Comment:
Does a_assets_dict certainly contain ac_uuid at this point, or can this throw?

Reply:
Good point. We protect the state calls above in this manner, we should also cover cases where ac_uuid isn't found in a_assets_dict. Will protect the whole loop with a catch as that's where the enumerate call is.